File: LeetCode_part/21.合并有序链表.py
class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # method1
        # if l1 is not None and l2 is not None:
        #     if l1.val > l2.val:
        #         l1, l2 = l2, l1  # l1指向较小值
        #     l1.next = self.mergeTwoLists(l1.next, l2) # next指向下一个较小值
        # return l1 or l2

        newList = ListNode(None)
        prev = newList
        while l1 is not None and l2 is not None:
            if l1.val <= l2.val:
                prev.next = l1
                l1 = l1.next
            else:
                prev.next = l2
                l2 = l2.next
            prev = prev.next

        prev.next = l1 if l1 is not None else l2  # None或者两者不为None的一个

        return newList.next

File: Book_part/剑指offer/21-合并有序链表.cpp
#include <iostream>
#include <vector>
#include <iterator>
#include <algorithm>
#include <stack>
#include <unordered_map>
#include <cmath>
#include <bitset>
#include <stdio.h>
#include <string.h>

using namespace std;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
    {
        if (l1 == nullptr)
            return l2;
        if (l2 == nullptr)
            return l1;

        ListNode *res = new ListNode(-1);
        ListNode *ptr = res;

        while (l1 != nullptr && l2 != nullptr)
        {
            if (l1->val <= l2->val)
            {
                ptr->next = l1;
                l1 = l1->next;
            }
            else
            {
                ptr->next = l2;
                l2 = l2->next;
            }
            ptr = ptr->next;
        }

        ptr->next = (l1 ? l1 : l2);
        return res->next;
    }

    ListNode *mergeTwoLists(ListNode *l1, ListNode *l2)
    {
        if (l1 == nullptr)
            return l2;
        if (l2 == nullptr)
            return l1;

        ListNode *res = new ListNode(-1);
        ListNode *ptr = res;

        merge(l1, l2, ptr);

        return res->next;
    }

    void merge(ListNode *l1, ListNode *l2, ListNode *ptr)
    {
        if (!l1 || !l2)
        {
            ptr->next = (l1 ? l1 : l2);
            return;
        }

        if (l1->val < l2->val)
        {
            ptr->next = l1;
            l1 = l1->next;
        }
        else
        {
            ptr->next = l2;
            l2 = l2->next;
        }

        ptr = ptr->next;
        merge(l1, l2, ptr);
    }
};